File: backend/src/platform/modrinth.ts
import fs from "node:fs";
import { WebSocket } from "ws";
import { Wfastdownload, Utils } from "../utils/utils.js";
import { modpack_info, XPlatform } from "./index.js";
import { join } from "node:path";

interface ModrinthManifest {
  files: Array<{ path: string; downloads: string[]; fileSize: number; }>;
  dependencies: {
    minecraft: string;
    forge: string;
    neoforge: string;
    "fabric-loader": string;
    [key: string]: string;
  };
}

export class Modrinth implements XPlatform {
  private utils: Utils;
  constructor() {
    this.utils = new Utils();
  }
  async getinfo(manifest: object): Promise<modpack_info> {
    let result: modpack_info = Object.create({});
    const local_manifest = manifest as ModrinthManifest;
    const depkey = Object.keys(local_manifest.dependencies);
    const loader = ["forge", "neoforge", "fabric-loader"];
    result.minecraft = local_manifest.dependencies.minecraft;
    for (let i = 0; i < depkey.length; i++) {
      const key = depkey[i];
      if (key !== "minecraft" && loader.includes(key)) {
        result.loader = key;
        result.loader_version = local_manifest.dependencies[key];
      }
    }
    return result;
  }
  async downloadfile(manifest: object,path:string,ws:WebSocket): Promise<void> {
    const index = manifest as ModrinthManifest;
    let tmp: [string, string][] = []
    index.files.forEach(async (e: { path: string; downloads: string[]; fileSize: number;}) => {
        if (e.path.endsWith(".zip")) {
            return;
          }
    const url = e.downloads[0].replace("https://cdn.modrinth.com",this.utils.modrinth_Durl)
    const unpath = join(path,e.path)
    tmp.push([e.downloads[0],unpath])
    // if (usemirror){
    //     tmp.push(["https://mod.mcimirror.top"+new URL(e.downloads[0]).pathname,unpath,String(e.fileSize)])
    // }else{
    //     tmp.push([e.downloads[0],unpath,String(e.fileSize)])
    // }
    });
    await Wfastdownload(tmp as unknown as [string, string],ws)
  }
}

File: backend/platform/modrinth.go
package platform

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"strings"

	"dex/backend/utils"
)

func NewModrinth() *Modrinth {
	return &Modrinth{
		urls: GetMirrorUrls(),
	}
}

type Modrinth struct {
	urls MirrorUrls
}

// modrinthManifest describes the structure of a Modrinth modrinth.index.json.
type modrinthManifest struct {
	Files []struct {
		Path      string   `json:"path"`
		Downloads []string `json:"downloads"`
		FileSize  int      `json:"fileSize"`
	} `json:"files"`
	Dependencies map[string]string `json:"dependencies"`
}

func (m *Modrinth) GetInfo(manifest map[string]any) (*ModpackInfo, error) {
	// Marshal back to JSON then parse into typed struct
	data, err := json.Marshal(manifest)
	if err != nil {
		return nil, fmt.Errorf("failed to re-marshal modrinth manifest: %w", err)
	}

	var mrManifest modrinthManifest
	if err := json.Unmarshal(data, &mrManifest); err != nil {
		return nil, fmt.Errorf("failed to parse modrinth manifest: %w", err)
	}

	result := &ModpackInfo{
		Minecraft: mrManifest.Dependencies["minecraft"],
	}

	// Find the mod loader from dependencies
	loaders := []string{"forge", "neoforge", "fabric-loader"}
	for _, key := range loaders {
		if _, ok := mrManifest.Dependencies[key]; ok {
			result.Loader = key
			result.LoaderVersion = mrManifest.Dependencies[key]
			break
		}
	}

	return result, nil
}

func (m *Modrinth) DownloadFile(manifest map[string]any, path string, progressFn func(total, completed int, name string)) error {
	// Parse manifest
	data, err := json.Marshal(manifest)
	if err != nil {
		return fmt.Errorf("failed to re-marshal modrinth manifest: %w", err)
	}

	var mrManifest modrinthManifest
	if err := json.Unmarshal(data, &mrManifest); err != nil {
		return fmt.Errorf("failed to parse modrinth manifest: %w", err)
	}

	// Build download list
	var downloadItems []utils.DownloadOption
	for _, e := range mrManifest.Files {
		// Skip .zip files
		if strings.HasSuffix(e.Path, ".zip") {
			continue
		}

		if len(e.Downloads) == 0 {
			continue
		}

		// Replace official CDN domain with mirror URL
		url := strings.Replace(e.Downloads[0], "https://cdn.modrinth.com", m.urls.ModrinthDurl, 1)
		unpath := filepath.Join(path, e.Path)

		downloadItems = append(downloadItems, utils.DownloadOption{
			URL:        url,
			FilePath:   unpath,
			UseChunked: true,
		})
	}

	if len(downloadItems) == 0 {
		return nil
	}

	// Use WFastDownload with progress reporting and chunked downloads enabled
	return utils.WFastDownload(downloadItems, progressFn)
}
